### PR TITLE
[LLDB][Minidump] Make workaround for the Dynamic loader issue

### PR DESCRIPTION
**Pull request by Jlalond**
In #119598 my recent TLS feature seems to break crashpad symbols. I have a few ideas on how this is happening, but for now as a mitigation I'm checking if the Minidump was LLDB generated, and if so leveraging the dynamic loader.

### CODE REVIEW

File: lldb/source/Plugins/Process/minidump/MinidumpParser.cpp [hunk 1]
```diff
@@ -45,6 +45,10 @@ llvm::ArrayRef<uint8_t> MinidumpParser::GetData() {
                                  m_data_sp->GetByteSize());
 }
 
+const llvm::minidump::Header *MinidumpParser::GetHeader() const {
+  return reinterpret_cast<llvm::minidump::Header *>(m_file.get());
```
Comment (labath) [inline lldb/source/Plugins/Process/minidump/MinidumpParser.cpp:49]: This looks wrong. I think you want m_file->header()

File: lldb/source/Plugins/Process/minidump/MinidumpParser.cpp [hunk 2]
```diff
@@ -453,10 +456,12 @@ MinidumpParser::FindMemoryRange(lldb::addr_t addr) {
 
   if (!GetStream(StreamType::Memory64List).empty()) {
     llvm::Error err = llvm::Error::success();
-    for (const auto &memory_desc :  GetMinidumpFile().getMemory64List(err)) {
-      if (memory_desc.first.StartOfMemoryRange <= addr 
-          && addr < memory_desc.first.StartOfMemoryRange + memory_desc.first.DataSize) {
-        return minidump::Range(memory_desc.first.StartOfMemoryRange, memory_desc.second);
+    for (const auto &memory_desc : GetMinidumpFile().getMemory64List(err)) {
```
Comment (labath) [inline lldb/source/Plugins/Process/minidump/MinidumpParser.cpp:459]: Please revert all the formatting changes to code you didn't modify. You can submit those as a separate PR if you want.

File: llvm/include/llvm/BinaryFormat/Minidump.h [hunk 3]
```diff
@@ -32,6 +32,10 @@ LLVM_ENABLE_BITMASK_ENUMS_IN_NAMESPACE();
 struct Header {
   static constexpr uint32_t MagicSignature = 0x504d444d; // PMDM
   static constexpr uint16_t MagicVersion = 0xa793;
+  // We set all the high bits flag to indicate this is from LLDB.
+  // We don't want to conflict with anything Microsoft is using the flags for
+  // and the highest bits are not currently being used.
+  static const uint32_t LLDB_HEADER_FLAG = 0xF0000000;
```
Comment (labath) [inline llvm/include/llvm/BinaryFormat/Minidump.h:38]: uint64_t to match the width of the field, and change the name to fit the convention of the constants above.

**Review by labath — COMMENTED** [review]
Marking a file as being generated by lldb makes sense, but given that there's no way (is there?) to reserve a flag bit for lldb, and the flag bits are a relatively scarce commodity (there's 64 of them total, how many are free right now?).

What would you say to creating a new "CreatedByLLDB" stream type and using that as the marker? There are 4 billion stream types, so the danger of collision is much lower, and we also have ample precedent for creating custom stream types. The thing I like about that is the we can then also insert additional metadata about the minidump into the stream, for example, whether it includes a linker rendezvous structures, which I think is the important part here (I'm not saying you have to do that now -- we can start out with a completely empty stream -- but I like how this gives us an option to do that in the future).

**Comment by Jlalond** [comment]
@labath I went with your suggestion as the better idea. I ended up just writing 'LLDB' to the stream, because the current API behavior for MinidumpParser is to return an empty ArrayRef. So there was no direct way to detect an existing but empty stream. For now I just set the check if it's `<= 4 bytes`. So in the future we can add extra LLDB centric information, and we just want to keep the Magic bytes to keep compatibility.

File: llvm/include/llvm/BinaryFormat/Minidump.h [hunk 4]
```diff
@@ -52,6 +52,7 @@ enum class StreamType : uint32_t {
 #include "llvm/BinaryFormat/MinidumpConstants.def"
   Unused = 0,
   LastReserved = 0x0000ffff,
+  LLDBGenerated = 0x4C4C4442, // ASCII for 'LLDB'
```
Comment (labath) [inline llvm/include/llvm/BinaryFormat/Minidump.h:55]: Please put this in llvm/BinaryFormat/MinidumpConstants.def. We already have breakpad and facebook types defined there, so I don't see a reason to do something different here.

File: lldb/source/Plugins/Process/minidump/ProcessMinidump.cpp [hunk 5]
```diff
@@ -354,6 +354,39 @@ DataExtractor ProcessMinidump::GetAuxvData() {
                        GetAddressByteSize(), GetAddressByteSize());
 }
 
+bool ProcessMinidump::IsLLDBMinidump() {
+  // If we've already checked, return the cached value
+  if (m_is_lldb_generated.has_value())
+    return *m_is_lldb_generated;
+
+  // If the minidump doesn't have a LLDBGeneratedStream, it's not an LLDB
```
Comment (labath) [inline lldb/source/Plugins/Process/minidump/ProcessMinidump.cpp:362]: There seems to be something missing in this sentence.

File: lldb/source/Plugins/Process/minidump/ProcessMinidump.cpp [hunk 6]
```diff
@@ -354,6 +354,39 @@ DataExtractor ProcessMinidump::GetAuxvData() {
...
+  // We also check to see if the section was generated correctly, but not
+  // enforcing an exact size so we can change it in the future without
+  // impacting older generated Minidumps.
+  llvm::ArrayRef<uint8_t> lldbStream =
+      m_minidump_parser->GetStream(StreamType::LLDBGenerated);
+  if (lldbStream.empty() || lldbStream.size() <= sizeof(StreamType)) {
```
Comment (labath) [inline lldb/source/Plugins/Process/minidump/ProcessMinidump.cpp:368]:
```suggestion
  if (lldbStream.size() < sizeof(StreamType)) {
```
`empty` is subsumed by the size check, and I think you need a strict comparison here.

File: lldb/source/Plugins/Process/minidump/ProcessMinidump.cpp [hunk 7]
```diff
@@ -354,6 +354,39 @@ DataExtractor ProcessMinidump::GetAuxvData() {
...
+    m_is_lldb_generated = false;
+    return false;
+  }
+
+  const uint32_t *lldbStreamType =
+      reinterpret_cast<const uint32_t *>(lldbStream.data());
```
Comment (labath) [inline lldb/source/Plugins/Process/minidump/ProcessMinidump.cpp:374]:
```suggestion
  uint32_t lldbStreamType = llvm::support::endian::read<uint32_t, llvm::endianness::little>(lldbStream.data());
```

We only support little-endian minidumps right now, but it'd be nice if they could be read on a system of any endianness. I'm also not sure if this is guaranteed to be aligned.

File: lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp [hunk 8]
```diff
@@ -126,6 +129,17 @@ Status MinidumpFileBuilder::AddDirectory(StreamType type,
   return error;
 }
 
+Status MinidumpFileBuilder::AddLLDBGeneratedStream() {
+  Status error;
+  StreamType type = StreamType::LLDBGenerated;
```
Comment (labath) [inline lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp:134]:
```suggestion
  llvm::support::ulittle32_t type = StreamType::LLDBGenerated;
```
.. for the same reason. (You may need to add some static_casts to get the initialization to compile)

**Review by labath — APPROVED** [review]
Thank you. If you really want to keep the magic header, then I guess it's fine, but I think it adds a lot of unnecessary code and is not particularly consistent with the other stream types (I don't remember seeing empty streams, but I also don't remember seeing streams that begin with a magic value -- it's kinda pointless since the stream type already serves as a sort of a magic value).

I don't find the "MinidumpParser does not distinguish empty/missing streams" argument particularly convincing, since the underlying llvm class (`MinidumpFile`) does distinguish them and a `MinidumpParser::ContainsStream()` function would be a lot simpler than this magic handling code.

File: lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp (hunk 8 above)
Comment (Jlalond) [reply lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp:134]: Dropped the magic bytes and added empty stream. Resolving

**Comment by Jlalond** [comment]
> I don't find the "MinidumpParser does not distinguish empty/missing streams" argument particularly convincing, since the underlying llvm class (`MinidumpFile`) does distinguish them and a `MinidumpParser::ContainsStream()` function would be a lot simpler than this magic handling code.

After looking at your review, I agree. I'm not sure why I didn't want to add more features to Minidump parser, but this is much much cleaner now.

File: lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp [hunk 9]
```diff
@@ -89,8 +89,11 @@ Status MinidumpFileBuilder::AddHeaderAndCalculateDirectories() {
         "Failed to fill in header and directory "
         "sections. Written / Expected (%" PRIx64 " / %" PRIx64 ")",
         new_offset, m_saved_data_size);
-
   return error;
+  if (error.Fail())
+    return error;
```
Comment (labath) [inline lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp:94]: Remove the extra return?

(It would be nice to at least have a test that verifies that the lldb stream is actually written)

**Review by labath — APPROVED** [review]

File: lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp (hunk 9 above)
Comment (Jlalond) [reply lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp]: We do have a [test that depends on the dynamic loader](https://github.com/llvm/llvm-project/blob/af83093933ca73bc82c33130f8bda9f1ae54aae2/lldb/test/API/functionalities/process_save_core_minidump/TestProcessSaveCoreMinidump.py#L611), but it was passing in CI, even with this incorrect return which is unexpected. I tested locally by capturing a minidump with TLS variables on an older version of lldb, and verified there is no dyld or TLS variables when loaded with LLDB built on this patch. Capturing and loading on this patch ensured it worked. This does mean we're making a breaking change for behavior, where all older TLS minidumps will no longer work.

Comment (labath) [reply lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp]: Ultimately I think this patch is a workaround for the fact that the dynamic loader (posix) and process (minidump) classes can't agree on how to load the modules. The dynamic loader expect that it will be completely in charge of module loading, and this works fine in case of a live process, as it always contains enough information to determine the set of loaded modules. A minidump does not (always) contains this information, but (maybe for that same reason) it contains a explicit list of loaded modules (which is handled by the process class). When we start to use both sources, things get messy. I'm not sure what's the best way to handle this situation, but I think this patch approximates the desired behavior relatively well, and I think it is useful to be able to tell who generated a minidump.

Comment (Jlalond) [reply lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp]: @labath I agree, I think it would be ideal to agree with Breakpad/Crashpad on a section about what program generated it, including version of that software. From Memory I know there is some brakepad relevant data, but a unified section would be great.

Comment (clayborg) [reply lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp]: We should change all dynamic loaders to always check with the process via:
```
lldb_private::StructuredData::ObjectSP Process::GetLoadedDynamicLibrariesInfos()
```
This allows processes to tell the dynamic loaders where things are loaded in case they know, like in the case of minidumps. So we can probably make the POSIX dynamic loader a bit smarter when it comes to dealing with core files.

Comment (clayborg) [reply lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp]: If nothing comes back from this call, then the dynamic loader will use its normal means to locate things, else, it should use the list that the process provides.

Comment (Jlalond) [reply lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp]: @clayborg Just so I follow, you're recommending we embed a structured data (really just Json) into a section? If so, I really like the proposal, because it allows us to have a fairly loose schema across Brakepad, Crashpad and LLDB.

Comment (clayborg) [reply lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp]: No, we should implement the `lldb_private::StructuredData::ObjectSP Process::GetLoadedDynamicLibrariesInfos()` virtual method in ProcessMinidump and return a structured data that describes where everything is loaded. Then have the POSIX dynamic loader call this method on the process object and see if it returns anything. If it does, then we use this as the truth, else we discover the shared libraries using the existing code in the POSIX dynamic loader.

Comment (clayborg) [reply lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp]: (since ProcessMinidump has the shared library load list embedded in a directory).

Comment (Jlalond) [reply lldb/source/Plugins/ObjectFile/Minidump/MinidumpFileBuilder.cpp]: @clayborg Got it, from the context of the conversation I thought you wanted that structure data embedded in the Corefile. 

I think that makes sense instead of using this Flag stream.
